Source/PINDiskCache.m

```diff
@@ -824,7 +833,10 @@ - (void)synchronouslyLockFileAccessWhileExecutingBlock:(PINCacheBlock)block
 
 - (BOOL)containsObjectForKey:(NSString *)key
 {
-    return ([self fileURLForKey:key updateFileModificationDate:NO] != nil);
+    if ([_knownKeys containsObject:key]) {
+        return ([self fileURLForKey:key updateFileModificationDate:NO] != nil);
```

Comment:
Can we skip this fileURLForKey: check, or maybe convert it into an assertion? If it is expected that this could fail, it would be good to add a comment here describing what case that occurs in.

Because the fileURL generation does show up in profiles, it seems like a shame to generate it here and then have to regenerate it in the calling code.

It might be possible to just ensure _knownKeys is kept up to date, and if an access is attempted and the file isn't present, then remove the key from _knownKeys (and throw an assertion there too?)

Reply:
knownKeys actually can't be guaranteed because the underlying directory is stored in a cache directory which iOS could empty at any time. Sadly I've found little official documentation on when this actually happens in practice, and I assume it doesn't occur while your app is running, but…

If it's not in knownKeys it's definitely not on disk, but if it is in knownKeys it might not be on disk.

Also of note, containsObjectForKey is not used internally by objectForKey (and in fact is not used internally at all) so this check doesn't happen twice in that case.